enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,731 @@
+the same, we simply will allow you to apply partitioning on non single node
+topologies.
+
+We will use the global identifier to correctly modify the pod spec with the
+`requests.cpu` for the new `requests[management.workload.openshift.io/cores]`
+that are used by the workload partitioning feature.
+
+However, for Single-Node we will continue to check the conventional way to be
+able to support the upgrade flow from 4.11 -> 4.12. After 4.12 release that
+logic should no longer be needed and will be removed.
+
+### Node Admission Plugin
+
+We want to ensure that nodes that are not setup for CPU Partitioning do not get
+added to a cluster that is designated for CPU Partitioning. We will create an
+admission plugin for Nodes that will validate if a cluster is setup for CPU
+Partitioning then the node MUST contain a Capacity resource for
+`workload.openshift.io/cores`. All node creation requests from Kubelet currently add
+that information on boot up when registering with the API Server, we will
+leverage that to ensure a CPU Partitioned cluster only contains nodes created
+for CPU Partitioning.
+
+We will also keep in mind upgrades from Single Node clusters which already
+contain this feature. During initial upgrade, Single Node clusters will not
+contain the CPUPartitioningMode, for this reason we will fall back to checking
+with the old logic for Single Node to insure we do not cause issues when
+upgrading. This check for Single Node should be something that happens on initial
+upgrades, as NTO will update the `Infra.Status.CPUPartitioningMode` to the correct
+value after initial boot.
+
+### Performance Profile Controller
+
+Currently workload partitioning involves configuring CRI-O and Kubelet earlier
+in the processes as a separate machine config manifest that requires the same
+information present in the `PerformanceProfile` resource, that being the
+`isolated` and `reserved` CPU sets. Because configuring multiple resources with
+the right CPU sets consistently is error prone, we want to extend the
+PerformanceProfile API to include settings for workload partitioning.
+
+When installation is done and workload partitioning is set then from that point
+on the `kubelet` and `crio` only need to be configured with the desired CPU set
+to use. We currently express this to customers via the `reserved` CPU set as
+part of the performance profile api.
+
+We want to add a new `workloads` field to the `cpu` field that contains a list
+of enums for defining which workloads to pin. This should allow us to expand
+this in the future if desired, but in this enhancement we will only support
+`Infrastructure` which defines all of the Openshift workloads.
+
+```yaml
+apiVersion: performance.openshift.io/v2
+kind: PerformanceProfile
+metadata:
+  name: openshift-node-workload-partitioning-custom
+spec:
+  cpu:
+    isolated: 2-3
+    reserved: 0,1
+    # New addition
+    workloads:
```

Comment:
Is there an implied relationship between "reserved" and "workloads" such that any item in "workloads" will run on the "reserved" CPUs? Looking at the structure that is hard to understand. Maybe they can be tied together such as a list of types and their CPU associations. The cpu would be either a list of cores or one of "isolated" or "reserved". Something like this:
```
    workloads:
      - type: management
        cpu: reserved
      - type: somethingelse
        cpu: 0-2
```

Reply:
These are great call outs Ian, I originally had something like this working but it started to become a little complicated to implement because it started to confuse the explicit intentions expressed by `reserved` and `isolated` in the PerformanceProfile. 

Since the PerformanceProfile defines the `reserved` set and `isolated` set and with the old method being that the `reserved` set was mirrored to our `crio` and `kubelet` config, the idea I gravitated towards was that we simplify the user experience if we consolidate that logic with the current PerforamnceProfile and allow the `workloads` enums to be that path of identifying which containers `crio` is configured to alter. So you're right that there is a new implication that what you're really saying as a user of this API is that:

For the targeted Nodes/Pools I want these `reserved` CPUs to be used for kubernetes and the `workloads` defined in this enum, and the `isolated` are only for my workloads.

Reply:
The api is confusing though. `workloads` here means the infrastructure pods, not customer workloads. It also lives in a section where all the other fields accept cpu sets.

I consider a higher level map (spec.workloadPinning[infrastructure]=reserved) similar to what Ian proposed to be much more intuitive.

Reply:
After our discussion, it was decided that for this enhancement we would be changing the PerformanceProfile API